crates/forge-runner/src/running/copied_code.rs

```diff
@@ -0,0 +1,384 @@
+    runner.run_from_entrypoint(
+        entry_point.pc(),
+        &args,
+        verify_secure,
+        Some(program_segment_size),
+        hint_processor,
+    )?;
+
+    maybe_fill_holes(entry_point, runner)?;
+
+    verify_secure_runner(runner, false, Some(program_segment_size))
+        .map_err(CairoRunError::VirtualMachine)?;
+
+    Ok(())
+}
+
+#[allow(clippy::items_after_statements, clippy::needless_pass_by_value)]
+/// Why copied: Required by `run_entry_point`
+///
+/// Fills the holes after running the entry point.
+/// Currently only fills the holes in the rc96 segment.
+fn maybe_fill_holes(
+    entry_point: EntryPointV1,
+    runner: &mut CairoRunner,
+) -> Result<(), EntryPointExecutionError> {
+    let Some(rc96_offset) = entry_point
+        .builtins
+        .iter()
+        .rev()
+        .position(|name| *name == BuiltinName::range_check96)
+    else {
+        return Ok(());
+    };
+    let rc96_builtin_runner = runner
+        .vm
+        .get_builtin_runners()
+        .iter()
+        .find_map(|builtin| {
+            if let BuiltinRunner::RangeCheck96(rc96_builtin_runner) = builtin {
+                Some(rc96_builtin_runner)
+            } else {
+                None
+            }
+        })
+        .expect("RangeCheck96 builtin runner not found.");
+
+    // 'EntryPointReturnValues' is returned after the implicits and its size is 5,
+    // So the last implicit is at offset 5 + 1.
+    const IMPLICITS_OFFSET: usize = 6;
+    let rc_96_stop_ptr = (runner.vm.get_ap() - (IMPLICITS_OFFSET + rc96_offset))
+        .map_err(|err| CairoRunError::VirtualMachine(VirtualMachineError::Math(err)))?;
+
+    let rc96_base = rc96_builtin_runner.base();
+    let rc96_segment: isize = rc96_base
+        .try_into()
+        .expect("Builtin segment index must fit in isize.");
+
+    let Relocatable {
+        segment_index: rc96_stop_segment,
+        offset: stop_offset,
+    } = runner
+        .vm
+        .get_relocatable(rc_96_stop_ptr)
+        .map_err(CairoRunError::MemoryError)?;
+    assert_eq!(rc96_stop_segment, rc96_segment);
+
+    // Update `segment_used_sizes` to include the holes.
+    runner
+        .vm
+        .segments
+        .segment_used_sizes
+        .as_mut()
+        .expect("Segments used sizes should be calculated at this point")[rc96_base] = stop_offset;
+
+    for offset in 0..stop_offset {
+        match runner.vm.insert_value(
+            Relocatable {
+                segment_index: rc96_segment,
+                offset,
+            },
+            Felt::zero(),
+        ) {
+            // If the value is already set, ignore the error.
+            Ok(()) | Err(MemoryError::InconsistentMemory(_)) => {}
+            Err(err) => panic!("Unexpected error when filling holes: {err}."),
+        }
+    }
+
+    Ok(())
+}
+
+// Reason copied: Private function
+pub(crate) fn prepare_program_extra_data(
+    runner: &mut CairoRunner,
+    // contract_class: &CompiledClassV1,
```

Comment:
Looks like it can be removed (?)

Reply:
The idea is to upstream as much of the code from this file as possible so don't want to modify it yet. Linked issue https://github.com/foundry-rs/starknet-foundry/issues/3293